chainer/training/updaters/multiprocess_parallel_updater.py

```diff
@@ -0,0 +1,223 @@
+                                 null_stream.ptr)
+                del gg
+                self.model.cleargrads()
+                if pp is None:
+                    pp = self.model.gather_params()
+                self.comm.bcast(pp.data.ptr, pp.size, nccl.NCCL_FLOAT, 0,
+                                null_stream.ptr)
+                self.model.scatter_params(pp)
+                pp = None
+
+                # Sending observation via pipe is too slow.
+                # self.pipe.send(observation)
+
+
+class MultiprocessParallelUpdater(StandardUpdater):
+    """Implementation of a multiprocess parallel GPU Updater.
+
+    This is an implementation of :class:`Updater` that uses multiple GPUs
+    with multi-process data parallelism. It uses Nvidia NCCL for communication
+    between multiple GPUs.
+
+    It behaves similarly to :class:`~chainer.training.StandardUpdater`. The
+    update routine is modified to support data-parallel computation on multiple
+    GPUs in one machine. It is based on synchronous parallel SGD: it
+    parallelizes the gradient computation over a mini-batch, and updates the
+    parameters only in the main device.
+
+    Args:
+        iterators: List of dataset iterator for the training dataset. The
+            number of the iterators must be same to the number of GPUs you use.
+        optimizer: Optimizer to update parameters. The model should be attached
+            to the optimizer.
+        converter: Converter function to build input arrays. Each batch
+            extracted by the iterator is split equally between the devices and
+            then passed with corresponding ``device`` option to this function.
+            :func:`~chainer.dataset.concat_examples` is used by default.
+        devices: Dictionary or list of devices to which the training data is
+            sent. The master device will be the first one in the list or the
+            value attached to the key ``'main'``.
+
+    """
+
+    def __init__(self, iterators, optimizer, converter=convert.concat_examples,
+                 devices=None):
+        if not MultiprocessParallelUpdater.available():
+            raise Exception(
+                'NCCL is not enabled. MultiprocessParallelUpdater '
+                'requires NCCL.\n'
+                'Please reinstall chainer after you install NCCL.\n'
+                '(see https://github.com/pfnet/chainer#installation).')
+
+        assert len(iterators) == len(devices)
+        for iterator in iterators[1:]:
+            assert len(iterator.dataset) == len(iterators[0].dataset)
+
+        # Correct optimizer parameters for new minibatch size
+        optim = optimizer.__class__.__name__
+        if optim in ('Adam', 'AdaGrad', 'RMSprop'):
+            optimizer.eps *= len(devices)
+        elif optim in ('RMSpropGraves', 'AdaDelta'):
+            optimizer.eps *= len(devices) ** 2  # not quite right for AdaDelta
+        else:
```

Comment:
This looks a little dangerous since newly added optimizer may not have `.lr` instance variable. How about replacing this with `elif optim in (<all the currently known optimizers>):` and then adding `else: Raise ...`? This may not be the best solution though.